### PR TITLE
Add support for saving client responses

### PR DESCRIPTION
**Pull request by tenyo**

### CODE REVIEW

File: actions/responses.go [hunk 1]
```diff
@@ -63,7 +66,23 @@ func ResponsesPost(c buffalo.Context) error {
...
+		return c.Error(500, errors.New("Error generating outcome id"))
+	}
 	log.Println("Outcome response:", outcome)
 
+	// save responses to S3
+	// we also include the original questions and the outcome that was returned to the client
+	s := struct {
+		GeneratedOutcome  Outcome
+		ClientResponse    models.Responses
+		OriginalQuestions models.Questions
```
Comment (fishnix) [inline actions/responses.go]: why not just call this `Outcome`, `Responses`, and `Questions`? 😆

Comment (tenyo) [reply actions/responses.go]: I thought that reads better in the json that gets generated.

File: actions/responses.go [hunk 2]
```diff
@@ -63,7 +66,23 @@ func ResponsesPost(c buffalo.Context) error {
...
+	}{outcome, responses, questions}
+	path := fmt.Sprint(responses.Path(c.Param("campaign")), outcome.ID.String(), ".json")
```
Comment (fishnix) [inline actions/responses.go]:
```golang
package test

import (
	"fmt"
	"strings"
	"testing"
)

var resp = "foobar"
var id = "xxxxx-yyyyyy-zzzzzz-99999"

func BenchmarkJoin(b *testing.B) {
	for i := 0; i < b.N; i++ {
		s := strings.Join([]string{resp, id, ".json"}, "")
		_ = s
	}
}

func BenchmarkSprintf(b *testing.B) {
	for i := 0; i < b.N; i++ {
		s := fmt.Sprintf("%s%s.json", resp, id)
		_ = s
	}
}

func BenchmarkSprint(b *testing.B) {
	for i := 0; i < b.N; i++ {
		s := fmt.Sprint(resp, id, ".json")
		_ = s
	}
}

func BenchmarkConcat(b *testing.B) {
	for i := 0; i < b.N; i++ {
		s := resp + id + ".json"
		_ = s
	}
}
```

```
➜  tmp go test -bench=. -benchtime=20s
goos: darwin
goarch: amd64
BenchmarkJoin-8      	500000000	        65.9 ns/op
BenchmarkSprintf-8   	200000000	       176 ns/op
BenchmarkSprint-8    	200000000	       189 ns/op
BenchmarkConcat-8    	500000000	        50.5 ns/op
PASS
ok  	_/Users/fish/tmp/go/yale/tmp	180.424s
```

File: actions/responses.go (hunk 1 above)
Comment (fishnix) [reply actions/responses.go]: you thought wrong

Comment (fishnix) [reply actions/responses.go]: 🤣

File: libs/s3/objects.go [hunk 3]
```diff
@@ -77,3 +78,39 @@ func (s Client) Load(i interface{}, path string) error {
 
 	return nil
 }
+
+// PutObject takes a byte array and saves it in S3 with the given key name
+func (s Client) PutObject(object []byte, key string, ct string) error {
```
Comment (fishnix) [inline libs/s3/objects.go]:
```golang
func (s Client) PutObject(object []byte, key, ct string) error {
```
probably use `contentType` as the variable name too - its more descriptive

Comment (fishnix) [reply libs/s3/objects.go]: it might be more intuitive to have the order changed too -- `ket, contentType, object`

File: models/responses.go [hunk 4]
```diff
@@ -5,4 +5,10 @@ type Responses struct {
 	List              map[string]string `json:"responses"`
 	RisklevelsVersion string            `json:"risklevels_version"`
 	QuestionsVersion  string            `json:"questions_version"`
+	Metadata          interface{}       `json:"metadata"`
+}
+
+// Path returns the main S3 path containing responses for a campaign
+func (r Responses) Path(c string) string {
+	return "responses/" + c + "/"
```
Comment (fishnix) [inline models/responses.go]: 🤷‍♂️

**Comment by fishnix** [comment]
looks good! 👍

File: models/responses.go (hunk 4 above)
Comment (tenyo) [reply models/responses.go]: The specific S3 path is embedded in the model and can be made configurable later.

**Comment by fishnix** [comment]
👍

**Pull request merged**